src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -70,15 +76,39 @@ internal sealed class VisualStudioProject
         /// </summary>
         private readonly FileChangeWatcher.IContext _fileReferenceChangeContext;
 
+        /// <summary>
+        /// track whether we have been subscribed to <see cref="IDynamicFileInfoProvider.Updated"/> event
+        /// </summary>
+        private readonly HashSet<IDynamicFileInfoProvider> _eventSubscriptionTracker = new HashSet<IDynamicFileInfoProvider>();
+
+        /// <summary>
+        /// map original dynamic file path to <see cref="DynamicFileInfo.FilePath"/>
+        /// 
+        /// original dyanmic file path points to something like xxx.cshtml that are given to project system
+        /// and <see cref="DynamicFileInfo.FilePath"/> points to a mapped file path provided by <see cref="IDynamicFileInfoProvider"/>
+        /// and how and what it got mapped to is up to the provider. 
+        /// 
+        /// Workspace will only knows about <see cref="DynamicFileInfo.FilePath"/> but not the original dynamic file path
+        /// </summary>
+        private readonly Dictionary<string, string> _dynamicFilePathMaps = new Dictionary<string, string>();
```

Comment:
if this is a file path, then i'm surprised we're not initializing with some appropriate equality comparer.  What't he contract on these file paths and case sensitivity?